pkg/traceql/ast.go

```diff
@@ -285,15 +287,22 @@ func (f SpansetFilter) evaluate(input []*Spanset) ([]*Spanset, error) {
 				continue
 			}
 
-			matchingSpans = append(matchingSpans, s)
+			f.matchingSpansBuffer = append(f.matchingSpansBuffer, s)
+		}
+
+		if len(f.matchingSpansBuffer) == 0 {
+			continue
 		}
 
-		if len(matchingSpans) == 0 {
+		if len(f.matchingSpansBuffer) == len(ss.Spans) {
+			// All matched, so we return the input as-is
+			// and preserve the local buffer.
+			output = append(output, ss)
 			continue
 		}
 
 		matchingSpanset := *ss
```

Comment:
should this use the `ss.clone()` method?

come to think of it, i wonder what the difference between clone and value assignment here is.

Reply:
I think this was written before clone() but they seem the same to me.  Clone could even be this one-liner instead of field-by-field copy.